src/nccl/result.rs

```diff
@@ -0,0 +1,343 @@
+}
+
+pub unsafe fn broadcast(
+    sendbuff: *const ::core::ffi::c_void,
+    recvbuff: *mut ::core::ffi::c_void,
+    count: usize,
+    datatype: sys::ncclDataType_t,
+    root: ::core::ffi::c_int,
+    comm: sys::ncclComm_t,
+    stream: sys::cudaStream_t,
+) -> Result<NcclStatus, NcclError> {
+    unsafe { sys::ncclBroadcast(sendbuff, recvbuff, count, datatype, root, comm, stream).result() }
+}
+
+pub unsafe fn all_reduce(
+    sendbuff: *const ::core::ffi::c_void,
+    recvbuff: *mut ::core::ffi::c_void,
+    count: usize,
+    datatype: sys::ncclDataType_t,
+    op: sys::ncclRedOp_t,
+    comm: sys::ncclComm_t,
+    stream: sys::cudaStream_t,
+) -> Result<NcclStatus, NcclError> {
+    unsafe { sys::ncclAllReduce(sendbuff, recvbuff, count, datatype, op, comm, stream).result() }
+}
+
+pub unsafe fn reduce_scatter(
+    sendbuff: *const ::core::ffi::c_void,
+    recvbuff: *mut ::core::ffi::c_void,
+    recvcount: usize,
+    datatype: sys::ncclDataType_t,
+    op: sys::ncclRedOp_t,
+    comm: sys::ncclComm_t,
+    stream: sys::cudaStream_t,
+) -> Result<NcclStatus, NcclError> {
+    unsafe {
+        sys::ncclReduceScatter(sendbuff, recvbuff, recvcount, datatype, op, comm, stream).result()
+    }
+}
+
+pub unsafe fn all_gather(
+    sendbuff: *const ::core::ffi::c_void,
+    recvbuff: *mut ::core::ffi::c_void,
+    sendcount: usize,
+    datatype: sys::ncclDataType_t,
+    comm: sys::ncclComm_t,
+    stream: sys::cudaStream_t,
+) -> Result<NcclStatus, NcclError> {
+    unsafe { sys::ncclAllGather(sendbuff, recvbuff, sendcount, datatype, comm, stream).result() }
+}
+
+pub unsafe fn send(
+    sendbuff: *const ::core::ffi::c_void,
+    count: usize,
+    datatype: sys::ncclDataType_t,
+    peer: ::core::ffi::c_int,
+    comm: sys::ncclComm_t,
+    stream: sys::cudaStream_t,
+) -> Result<NcclStatus, NcclError> {
+    unsafe { sys::ncclSend(sendbuff, count, datatype, peer, comm, stream).result() }
+}
+
+pub unsafe fn recv(
+    recvbuff: *mut ::core::ffi::c_void,
+    count: usize,
+    datatype: sys::ncclDataType_t,
+    peer: ::core::ffi::c_int,
+    comm: sys::ncclComm_t,
+    stream: sys::cudaStream_t,
+) -> Result<NcclStatus, NcclError> {
+    unsafe { sys::ncclRecv(recvbuff, count, datatype, peer, comm, stream).result() }
+}
+pub unsafe fn group_end() -> Result<NcclStatus, NcclError> {
+    unsafe { sys::ncclGroupEnd().result() }
+}
+
+pub unsafe fn group_start() -> Result<NcclStatus, NcclError> {
+    unsafe { sys::ncclGroupStart().result() }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::driver::CudaDevice;
+    use std::ffi::c_void;
+
+    #[test]
+    fn single_thread() {
+        let n_devices = CudaDevice::count().unwrap() as usize;
+        let n = 2;
+
+        let mut devs = vec![];
+        let mut sendslices = vec![];
+        let mut recvslices = vec![];
+        for i in 0..n_devices {
+            let dev = CudaDevice::new(i).unwrap();
```

Comment:
If you create the new device outside the loop can you call bind_to_thread() in this loop and below in the second loop?